endpoints/put.go

```diff
@@ -32,105 +31,70 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
 			return
 		}
 		defer r.Body.Close()
 
-		put := putAnyRequestPool.Get().(*PutRequest)
-		defer putAnyRequestPool.Put(put)
+		backend := &backendCallObject{client: backendClient, allowKeys: allowKeys}
 
-		err = json.Unmarshal(body, put)
+		backend.put = putAnyRequestPool.Get().(*PutRequest)
+		defer putAnyRequestPool.Put(backend.put)
+
+		err = json.Unmarshal(body, backend.put)
 		if err != nil {
 			http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
 			return
 		}
 
-		if len(put.Puts) > maxNumValues {
-			http.Error(w, fmt.Sprintf("More keys than allowed: %d", maxNumValues), http.StatusBadRequest)
-			return
-		}
-
-		resps := putResponsePool.Get().(*PutResponse)
-		resps.Responses = make([]PutResponseObject, len(put.Puts))
-		defer putResponsePool.Put(resps)
-
-		for i, p := range put.Puts {
-			if len(p.Value) == 0 {
-				http.Error(w, "Missing value.", http.StatusBadRequest)
-				return
-			}
-			if p.TTLSeconds < 0 {
-				http.Error(w, fmt.Sprintf("request.puts[%d].ttlseconds must not be negative.", p.TTLSeconds), http.StatusBadRequest)
-				return
-			}
-
-			var toCache string
-			if p.Type == backends.XML_PREFIX {
-				if p.Value[0] != byte('"') || p.Value[len(p.Value)-1] != byte('"') {
-					http.Error(w, fmt.Sprintf("XML messages must have a String value. Found %v", p.Value), http.StatusBadRequest)
-					return
-				}
-
-				// Be careful about the the cross-script escaping issues here. JSON requires quotation marks to be escaped,
-				// for example... so we'll need to un-escape it before we consider it to be XML content.
-				var interpreted string
-				json.Unmarshal(p.Value, &interpreted)
-				toCache = p.Type + interpreted
-			} else if p.Type == backends.JSON_PREFIX {
-				toCache = p.Type + string(p.Value)
-			} else {
-				http.Error(w, fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found %v", p.Type), http.StatusBadRequest)
-				return
-			}
-
-			logrus.Debugf("Storing value: %s", toCache)
-			u2, err := uuid.NewV4()
-			if err != nil {
-				http.Error(w, fmt.Sprintf("Error generating version 4 UUID"), http.StatusInternalServerError)
-			}
-			resps.Responses[i].UUID = u2.String()
-
-			ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
-			defer cancel()
-			// Only allow setting a provided key if configured (and ensure a key is provided).
-			if allowKeys && len(p.Key) > 0 {
-				s, err := backend.Get(ctx, p.Key)
-				if err != nil || len(s) == 0 {
-					resps.Responses[i].UUID = p.Key
+		// Get a response object from the resource pool that we'll fill with processed info
+		backend.resp = putResponsePool.Get().(*PutResponse)
+		backend.resp.Responses = make([]PutResponseObject, len(backend.put.Puts))
+		backend.resp.toCacheStrings = make([]string, len(backend.put.Puts))
+		defer putResponsePool.Put(backend.resp)
+
+		var cancel func()
+		backend.ctx, cancel = context.WithTimeout(context.Background(), 500*time.Millisecond)
```

Comment:
Doesn't this create a context that expires 500 milliseconds after midnight on Jan 1, 1970?

Reply:
I don't see why it would. WithTimeout takes a duration, not a timestamp.